File: C#/第二套/11900-1060308/11900-1060308/Services/DataService.cs
namespace FractionArithmetic.Services;

using FractionArithmetic.Models;
using System.IO;

public interface IDataService
{
    Task<IEnumerable<(Fraction f1, string op, Fraction f2)>> ReadDataAsync(string filePath);
}

public class DataService : IDataService
{
    public async Task<IEnumerable<(Fraction f1, string op, Fraction f2)>> ReadDataAsync(string filePath)
    {
        if (!File.Exists(filePath)) return Enumerable.Empty<(Fraction, string, Fraction)>();

        var lines = await File.ReadAllLinesAsync(filePath);
        return lines
            .Where(line => !string.IsNullOrWhiteSpace(line))
            .Select(line =>
            {
                var parts = line.Split(',', StringSplitOptions.TrimEntries);
                if (parts.Length >= 5)
                {
                    if (long.TryParse(parts[0], out long n1) &&
                        long.TryParse(parts[1], out long d1) &&
                        long.TryParse(parts[3], out long n2) &&
                        long.TryParse(parts[4], out long d2))
                    {
                        return (new Fraction(n1, d1), parts[2], new Fraction(n2, d2));
                    }
                }
                return default;
            })
            .Where(x => x != default);
    }
}

File: VB/第二套/11900-1060308/Services/DataService.vb
Imports System.IO

Namespace Services
    Public Interface IDataService
        Function LoadCalculations(path As String) As IEnumerable(Of CalculationInput)
    End Interface

    Public Class CalculationInput
        Public Property F1 As Models.Fraction
        Public Property Op As String
        Public Property F2 As Models.Fraction
    End Class

    Public Class DataService
        Implements IDataService

        Public Function LoadCalculations(path As String) As IEnumerable(Of CalculationInput) Implements IDataService.LoadCalculations
            Dim list As New List(Of CalculationInput)()
            If File.Exists(path) Then
                For Each line In File.ReadAllLines(path)
                    Dim ss = line.Split(","c)
                    If ss.Length >= 5 Then
                        Dim input As New CalculationInput()
                        input.F1 = New Models.Fraction(Val(ss(0)), Val(ss(1)))
                        input.Op = ss(2).Trim()
                        input.F2 = New Models.Fraction(Val(ss(3)), Val(ss(4)))
                        list.Add(input)
                    End If
                Next
            End If
            Return list
        End Function
    End Class
End Namespace
